### PR TITLE
Proper implement isSameNode method

### LINKED ISSUE #1325: Call to doc() inside XSL fails.
### What is the problem
`<xsl:value-of select="$lutdoc//product/id"/>` fails :
```
exerr:ERROR Exception while transforming node: Internal error evaluating
template in module [at line 36, column 16]
```

### What did you expect
When uncommenting the 2 lines in the provided test it should report :
`<ok> Lutdoc: First 100 Second 200 LutIds:100 200 Name:First</ok>`

### Describe how to reproduce or add a test
A 'self containing' test, a test that can be run without dependancies. 
Given an XML with only <id> and "A-Lut.xml" containing both <id> and <name> used as Look Up Table the XSL matches <id> in current input and looks up using doc("A-Lut.xml") to output <name>.
The "A-Lut.xml" is first stored in the base by "A-Bug.xql" to make the test 'self contained'.
The XML input and the XSL are built in memory by the test, then passed to transform:transform()

`A-Bug.xql` stored in /db/apps/bugs
```xquery
xquery version "3.0";
declare namespace xmldb = "http://exist-db.org/xquery/xmldb";
declare option exist:serialize "method=xml media-type=text/html";

let $lut :=
<lut>
   <product>
     <name>First</name>
     <id>100</id>
   </product>
   <product>
     <name>Second</name>
     <id>200</id>
   </product>
</lut>

let $store := xmldb:store('/db/apps/bugs', 'A-Lut.xml', $lut)

let $xmlin :=
<root>
   <product>
     <id>100</id>
   </product>
</root>

let $xslSheet :=
   <xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
version="2.0">
     <xsl:param name="lut">A-Lut.xml</xsl:param>
     <xsl:template match="id">
       <xsl:variable name="lutdoc" select="doc($lut)"/>
       Lutdoc:<xsl:value-of select="$lutdoc"/>
<!--      LutIds:<xsl:value-of select="$lutdoc//product/id"/> -->
<!--      Name:<xsl:value-of select="$lutdoc//product/id[. = current()]/../name"/>-->
     </xsl:template>
   </xsl:stylesheet>

let $xmlout := transform:transform($xmlin, $xslSheet, ())
return <ok>{$xmlout}</ok>
```

### Context information
With
- eXist-db version: 3.0 / acd0c14
- Java: 1.8.0_112
- Operating system: macOS 10.12.3 (16D32)

It fails on `<xsl:value-of select="$lutdoc"/>`
```
exerr:ERROR Exception while transforming node: I/O error reported by XML
parser processing file:/Users/remi/git/exist-develop/A-Lut.xml:
/Users/remi/git/exist-develop/A-Lut.xml (No such file or directory) [at
line 38, column 16]
```
doc($lut) seems to read from the file system not the database.

With patch https://github.com/eXist-db/exist/pull/1316
in eXide, it finds the file "A-Lut.xml" in the base and reports:
`<ok> Lutdoc:First100Second200</ok>`
doc($lut) finds the file in the base, but seems to get only text.

Sadly the patch corrects only part of the bug.
Uncomment <xsl:value-of select="$lutdoc//product/id"/> and it fails :
```
exerr:ERROR Exception while transforming node: Internal error evaluating
template in module [at line 38, column 16]
```

Uncomment the next line, tweak xsl param, and it reports correctly :
`<ok> Lutdoc: First 100 Second 200 LutIds:100 200 Name:First</ok>`
exactly as Saxon on command line (in fact this is the return from embedded Saxon)

For the sake of completeness by "tweak xsl param" I mean
replace `xsl:param name="lut">A-Lut.xml</xsl:param>`
with `<xsl:param name="lut">xmldb:exist:///db/apps/bugs/A-Lut.xml</xsl:param>`
witch is a way to bypass the bug.

Rémi

### LINKED ISSUE #1337: Getting "ERROR isSameNode(Node other) not implemented on other class org.exist.dom.persistent.DocumentImpl"
This is something that was working with eXist-db version 3.0RC2 and doesn't work anymore with 3.1.0.

When executing an XQuery, from eXide for instance, there is an issue with the instruction : 

```xquery
transform:transform(
 doc($xmlFileUri), 
 doc(concat($xsltPath, '/', 'FancyTreeVisuDeclB30.xsl')), $params
)
```

We get this in return : 
`exerr:ERROR Exception while transforming node: isSameNode(Node other) not implemented on other class org.exist.dom.persistent.DocumentImpl`

When debugging an execution of our application embeddind eXist, we have seen that this message is coming from the method is SameNode of class ElementImpl : 
```java
    @Override
    public boolean isSameNode(final Node other) {
        // This function is used by Saxon in some circumstances, and this partial implementation is required for proper Saxon operation.
        if(other instanceof IStoredNode) {
            return (this.nodeId == ((IStoredNode<?>) other).getNodeId() &&
                this.ownerDocument.getDocId() == ((IStoredNode<? extends IStoredNode>) other).getOwnerDocument().getDocId());
        }
        throw new DOMException(DOMException.NOT_SUPPORTED_ERR,
            "isSameNode(Node other) not implemented on other class " + other.getClass().getName());
    }
```

So it seems that DocumentImpl should implement IStoredNode...

### Context information
[eXist Version : 3.1.0] 
[eXist Build : 20170307214949] 
[Git commit : c679cd8] 
[Operating System : Windows Server 2008 R2 6.1 amd64]

### PR DESCRIPTION
**Pull request by shabanovd**
close #1325
close #1337

Attention: merge only after #1316

### CODE REVIEW

File: src/org/exist/http/servlets/XSLTServlet.java [hunk 1]
```diff
@@ -124,12 +119,11 @@ private boolean isCaching() {
         return caching;
     }
 
-
     @Override
     protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
 
-        final String stylesheet = (String) request.getAttribute(REQ_ATTRIBUTE_STYLESHEET);
-        if (stylesheet == null) {
+        final String uri = (String) request.getAttribute(REQ_ATTRIBUTE_STYLESHEET);
```
Comment (dizzzz) [inline src/org/exist/http/servlets/XSLTServlet.java]: probably this should  not be part of the PR ; please keep functional changes and refactoring a bit separated.

**Comment by dizzzz** [comment]
ah! it is a mix of two Mrs. this makes review a bit complex.... and much more than needed...

**Comment by shabanovd** [comment]
@dizzzz merge 1st PR and review this one ;-)

1st PR 8 days old already -(

File: test/src/org/exist/xquery/TransformTest.java [hunk 2]
```diff
@@ -49,8 +49,54 @@ public void transform() throws XMLDBException {
                 "<p>End Template 1</p>" +
                 "</doc>");
     }
-    
-    
+
+    @Test
+    public void isSameNodeTest() throws XMLDBException {
```
Comment (dizzzz) [inline test/src/org/exist/xquery/TransformTest.java]: what is the purpose of the test? is the name ok?

**Comment by dizzzz** [comment]
1st PR is not pulled in.... that is bad, will check why this is....

File: test/src/org/exist/xquery/TransformTest.java (hunk 2 above)
Comment (shabanovd) [inline test/src/org/exist/xquery/TransformTest.java]: Yeap, that test for [issue](https://github.com/eXist-db/exist/issues/1325). The problem at `isSameNode` method and that's why test have it name.

Comment (dizzzz) [inline test/src/org/exist/xquery/TransformTest.java]: did Remy confirm the fix?

Comment (shabanovd) [inline test/src/org/exist/xquery/TransformTest.java]: @dizzzz you are too fast, that PR one hour old. You should give @RemiKoutcherawy a bit more time ;-)
